ambry-store/src/test/java/com/github/ambry/store/BlobStoreTest.java

```diff
@@ -714,6 +714,62 @@ public Void call() throws Exception {
     }
   }
 
+  /**
+   * Test the case where a Put happens while a Delete is doing the preliminary check.
+   * Since the delete happens before put, delete should return a ID_NOT_FOUND error.
+   * @throws Exception
+   */
+  @Test
+  public void concurrentDeleteAndPutTest() throws Exception {
+    MockId id = getUniqueId();
+    final CountDownLatch getEndOffsetLatch = new CountDownLatch(1);
+    final CountDownLatch findKeyLatch = new CountDownLatch(1);
+    ((MockBlobStore) store).setBetweenGetEndOffsetAndFindKeyFunc(() -> {
+      getEndOffsetLatch.countDown();
+      try {
+        findKeyLatch.await();
+      } catch (Exception e) {
+      }
+    });
+
+    ExecutorService executorService = Executors.newFixedThreadPool(2);
+    try {
+      Future<Void> deleteFuture = executorService.submit(new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          MessageInfo info =
+              new MessageInfo(id, DELETE_RECORD_SIZE, id.getAccountId(), id.getContainerId(), time.milliseconds(),
+                  MessageInfo.LIFE_VERSION_FROM_FRONTEND);
+          store.delete(Collections.singletonList(info));
+          return null;
+        }
+      });
+      // Now make sure delete is on the way
+      getEndOffsetLatch.await();
+      Future<Void> putFuture = executorService.submit(new Callable<Void>() {
+        @Override
+        public Void call() throws Exception {
+          putOneBlobWithId(id, PUT_RECORD_SIZE, Utils.Infinite_Time);
+          return null;
+        }
+      });
+      putFuture.get();
```

Comment:
I feel like it's more appropriate to put `getEndOffsetLatch.await()` and `findKeyLatch.countDown()` into the `call()` method that performs `putOneBlobWithId `

Reply:
good idea.